Email Template Not Replacing Contact/Lead/Target Name Variable in Leads

<!--- Provide a general summary of the issue in the **Title** above -->
<!--- Before you open an issue, please check if a similar issue already exists or has been closed before. --->
<!--- If you have discovered a security risk please report it by emailing security@suitecrm.com. This will be delivered to the product team who handle security issues. Please don't disclose security bugs publicly until they have been handled by the security team. --->

#### Issue
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
I do see there are other issues with emails already listed #3829 #3321 #3310 #6462 but don't seem to be exactly the same but may be related.

When creating an email template and then inserting Contact/Lead/Target variable for $contact_first_name, for example if you send it to a lead it does not work (it worked in previous versions) it just leaves the variable there.  However, this works fine if you send it to a contact and replaces the variable with the first name no problem.  If I use the variable $lead_first_name and send it to a lead it work no problem.  However, using the Contact/Lead/Target variable is supposed to grab the right first name regardless of the record.

I should mention, it works as part of a workflow and subs in the right variable, just when you select a single lead to send an email to, it doesn't populate the value.

#### Expected Behavior
<!--- Tell us what should happen -->
Using the Contact/Lead/Target variable is supposed to grab the right first name regardless of the record when used in a template
#### Actual Behavior
<!--- Tell us what happens instead -->
<!--- Also please check relevant logs (suitecrm.log, php error.log etc.) -->
When creating an email template and then inserting Contact/Lead/Target variable for $contact_first_name, for example if you send it to a lead it does not work (it worked in previous versions) it just leaves the variable there as $contact_first_name when sending to the client.

I've checked error.log and there are no errors there to report.

#### Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->

#### Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug include code to reproduce, if relevant -->
1. Create an email template 
2. Add field Contact/Lead/Target (any field) $contact_first_name for example
3. Send the email to a Lead (and relate the email to the lead before sending)
4. Doesn't populate the lead's first name in email template

#### Context
<!--- How has this bug affected you? What were you trying to accomplish? -->
<!--- If you feel this should be a low/medium/high priority then please state so -->
This should be a high priority.  Sending out emails to people that you can't personalize kind of defeats the purpose of a CRM!  Looks like there are lots of bugs around sending emails and variables.  Would be great to get this fixed.

#### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* SuiteCRM Version used:  7.10.10
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)): Chrome 70.0.3538.102
* Environment name and version (e.g. MySQL, PHP 7):  PHP 7.1
* Operating System and version (e.g Ubuntu 16.04): Apache

I see this is closed but my campaigns AND emails still have the same problem as described by @pstevens71: "When creating an email template and then inserting Contact/Lead/Target variable for $contact_first_name, for example if you send it to a lead it does not work (it worked in previous versions) it just leaves the variable there as $contact_first_name when sending to the client."

This error in this log MAY or MAY NOT be related. The email is successfully sent but the email variable doesn't work.

I set the log to debug and this is where it seems to go fubar..... (but there are many other entries):

Thu Feb 21 15:24:55 2019 [12015][1][ERROR] Saving Email with invalid From name and/or Address. Details:
From field is not set.
From-name is not set.
From address and name pair is empty.
codes:1, 7, 14
Thu Feb 21 15:24:55 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:55 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:55 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:55 2019 [12015][1][ERROR] Invalid email from address or name detected before sending. Details:
From field is not set.
From-name is not set.
From address and name pair is empty.
codes:1, 7, 14

Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Saving Email with invalid From name and/or Address. Details:
From field is not set.
From address is not set.
From-name is not set.
From name is not set.
From address and name pair is empty.
codes:1, 4, 7, 10, 14
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Saving Email with invalid From name and/or Address. Details:
From field is not set.
From address is not set.
From-name is not set.
From name is not set.
From address and name pair is empty.
codes:1, 4, 7, 10, 14
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] fromUser: Conversion of from user format Y-m-d H:i failed
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Saving Email with invalid From name and/or Address. Details:
From field is not set.
From-name is not set.
From address and name pair is empty.
codes:1, 7, 14
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Unable to find relationship emails_email_templates
Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Array
(
[0] => Unhandled email save and store as sent error: 1
[1] => 6
)

Thu Feb 21 15:24:57 2019 [12015][1][ERROR] Array
(
[0] => Unhandled non gmail sent folder hander error: 2
[1] => 100
)



* SuiteCRM Version used:  7.11.2 (it didn't work on 7.11.1 either)
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)): Chrome 70.0.3538.102
* Environment name and version (e.g. MySQL, PHP 7):  PHP 7.2
* Operating System and version (e.g Ubuntu 16.04): Apache

--- a/modules/EmailTemplates/EmailTemplateParser.php
+++ b/modules/EmailTemplates/EmailTemplateParser.php
@@ -174,6 +174,18 @@ class EmailTemplateParser
 
         $parts = explode($charUnderscore, ltrim($variable, $charVariable));
         list($moduleName, $attribute) = [array_shift($parts), join($charUnderscore, $parts)];
+
+        /* Leads/Prospects/Users have a special variable naming scheme.
+        $contact_xxx for leads/prospects and $contact_user_xxx for users */
+        if (strtolower($moduleName) === 'contact') {
+            if (in_array($this->module->object_name, ['Lead', 'Prospect'], true)) {
+                $moduleName = strtolower($this->module->object_name);
+            } else if ($this->module->object_name == 'User' && str_begin(strtolower($attribute), 'user_')) {
+                $attribute = explode('_', $attribute, 2)[1];
+                $moduleName = 'user';
+            }
+        }
+
         if (in_array($attribute, static::$allowedVariables, true)) {
             return $this->getNonDBVariableValue($attribute);
         }

--- a/tests/unit/phpunit/modules/EmailTemplates/EmailTemplateTest.php
+++ b/tests/unit/phpunit/modules/EmailTemplates/EmailTemplateTest.php
@@ -1,5 +1,7 @@
 <?php
 
+require_once 'modules/EmailTemplates/EmailTemplateParser.php';
+
 class EmailTemplateTest extends SuiteCRM\StateCheckerPHPUnitTestCaseAbstract
 {
     public function setUp()
@@ -9,6 +11,42 @@ class EmailTemplateTest extends SuiteCRM\StateCheckerPHPUnitTestCaseAbstract
         global $current_user;
         get_sugar_config_defaults();
         $current_user = new User();
+    }
+
+    public function testEmailTemplateParser()
+    {
+        $emailTemplate = new EmailTemplate();
+        $emailTemplate->body_html = to_html('<h1>Hello $contact_name</h1>');
+        $emailTemplate->body = 'Hello $contact_name';
+        $emailTemplate->subject = 'Hello $contact_name';
+        $campaign = new Campaign();
+
+        $related = [new Lead(), new Contact(), new Prospect()];
+        foreach ($related as $bean) {
+            $bean->name = 'foobar';
+
+            $parser = new EmailTemplateParser($emailTemplate, $campaign, $bean, "", "");
+            $result = $parser->parseVariables();
+            $this->assertEquals('<h1>Hello foobar</h1>', from_html($result['body_html']));
+            $this->assertEquals('Hello foobar', $result['body']);
+            $this->assertEquals('Hello foobar', $result['subject']);
+        }
+    }
+
+    public function testEmailTemplateParserUser()
+    {
+        $emailTemplate = new EmailTemplate();
+        $emailTemplate->body = 'Hello $contact_user_full_name';
+        $campaign = new Campaign();
+
+        $bean = new User();
+        $bean->first_name = 'foo';
+        $bean->last_name = 'bar';
+        $bean->fill_in_additional_detail_fields();
+
+        $parser = new EmailTemplateParser($emailTemplate, $campaign, $bean, "", "");
+        $result = $parser->parseVariables();
+        $this->assertEquals('Hello foo bar', $result['body']);
     }
 
     public function testcreateCopyTemplate()

